CDI programmatic lookup - detect "unused" false positives at runtime

mkouba opened the pull request:
- previously, we only detected this problem and logged a warning when
obtaining a bean instance via BeanContainer (extensions)
- now we also cover CDI.current() and injected
javax.enterprise.inject.Instance
- the warning message was improved - include much more information
- added /quarkus/arc/removed-beans debug endpoint in the development
mode

mkouba committed:
CDI programmatic lookup - detect "unused" false positives at runtime

- previously, we only detected this problem and logged a warning when
obtaining a bean instance via BeanContainer (extensions)
- now we also cover CDI.current() and injected
javax.enterprise.inject.Instance
- the warning message was improved - include much more information
- added /quarkus/arc/removed-beans debug endpoint in the development
mode

mkouba commented:
The warning message looks like:
```
2020-09-03 12:14:13,373 WARN  [io.qua.arc.impl] (main) 
================================================================================
CDI: programmatic lookup problem detected
-----------------------------------------
At least one bean matched the required type and qualifiers but was marked as unused and removed during build
Removed beans:
	- PRODUCER_METHOD bean io.quarkus.arc.test.unused.UnusedExclusionTest$GamaProducer#ping() [types=[class io.quarkus.arc.test.unused.UnusedExclusionTest$Gama], qualifiers=[@javax.enterprise.inject.Default(), @javax.enterprise.inject.Any()]]
	- CLASS bean io.quarkus.arc.test.unused.UnusedExclusionTest$Gama [types=[class io.quarkus.arc.test.unused.UnusedExclusionTest$Gama], qualifiers=[@javax.enterprise.inject.Default(), @javax.enterprise.inject.Any()]]
Required type: class io.quarkus.arc.test.unused.UnusedExclusionTest$Gama
Required qualifiers: [@javax.enterprise.inject.Default()]
Solutions:
	- Application developers can eliminate false positives via the @Unremovable annotation
	- Extensions can eliminate false positives via build items, e.g. using the UnremovableBeanBuildItem
	- See also https://quarkus.io/guides/cdi-reference#remove_unused_beans
================================================================================
```

geoand commented on extensions/arc/deployment/src/test/java/io/quarkus/arc/test/unused/UnusedExclusionTest.java:
```diff
@@ -21,13 +35,56 @@
+                    @Override
+                    public void execute(BuildContext context) {
+                        BeanContainer beanContainer = context.consume(BeanContainerBuildItem.class).getValue();
+                        BytecodeRecorderImpl bytecodeRecorder = new BytecodeRecorderImpl(true,
+                                TestRecorder.class.getSimpleName(),
+                                "test", "" + TestRecorder.class.hashCode());
+                        // We need to use reflection due to some class loading problems
+                        Object recorderProxy = bytecodeRecorder.getRecordingProxy(TestRecorder.class);
```
Let's keep an eye out for future uses of this and try to move it into some test utility if it starts shows up

mkouba replied:
FYI @michalszynkiewicz

geoand commented:
This is super useful!

Does it make sense maybe to only have this detection enabled in dev and test mode? My thinking is that when running the regular jar it's too late to do anything about it and also to reduce memory usage

mkouba commented:
> My thinking is that when running the regular jar it's too late to do anything about it and also to reduce memory usage

It's never too late to do anything about it ;-) but I get your point. I think that additional memory consumption should be negligible for most apps. Note that even now we record all bean types of removable beans in the bytecode. Maybe we could make it configurable (and enabled by default)?

geoand commented:
> > My thinking is that when running the regular jar it's too late to do anything about it and also to reduce memory usage
> 
> It's never too late to do anything about it ;-) but I get your point. I think that additional memory consumption should be negligible for most apps. Note that even now we record all bean types of removable beans in the bytecode. Maybe we could make it configurable (and enabled by default)?

Sounds good to me :)

geoand commented on extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcConfig.java:
```diff
@@ -146,6 +146,15 @@
+     * @see ArcConfig#removeUnusedBeans
+     */
+    @ConfigItem(defaultValue = "true")
+    public boolean detectUnusedFalsePositives;
```
:+1:

geoand commented:
LGTM, but I think this is one for @manovotn to review more thoroughly :)

manovotn reviewed (approved):
That's really nice QoL improvement!

gastaldi commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/devmode/ArcEndpointProcessor.java:
```diff
@@ -13,15 +17,38 @@
     @Record(ExecutionTime.RUNTIME_INIT)
     @BuildStep(onlyIf = IsDevelopment.class)
-    void registerRoutes(HttpBuildTimeConfig config, ArcEndpointRecorder recorder, BuildProducer<RouteBuildItem> routes,
+    void registerRoutes(HttpBuildTimeConfig httConfig, ArcConfig arcConfig, ArcEndpointRecorder recorder,
```
```suggestion
    void registerRoutes(HttpBuildTimeConfig httpConfig, ArcConfig arcConfig, ArcEndpointRecorder recorder,
```

geoand commented:
CI seems to be consistently failing with this change:

```
2020-09-03T16:37:40.7744267Z [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.474 s <<< FAILURE! - in io.quarkus.vault.VaultSysITCase
2020-09-03T16:37:40.7750556Z [ERROR] io.quarkus.vault.VaultSysITCase  Time elapsed: 0.474 s  <<< ERROR!
2020-09-03T16:37:40.7784835Z java.lang.ExceptionInInitializerError
2020-09-03T16:37:40.7785178Z 	at java.base/java.lang.Class.forName0(Native Method)
2020-09-03T16:37:40.7785525Z 	at java.base/java.lang.Class.forName(Class.java:398)
2020-09-03T16:37:40.7785846Z 	at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:196)
2020-09-03T16:37:40.7786165Z 	at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:406)
2020-09-03T16:37:40.7786517Z 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$7(ClassBasedTestDescriptor.java:359)
2020-09-03T16:37:40.7786853Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-09-03T16:37:40.7787186Z 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:359)
2020-09-03T16:37:40.7787525Z 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:189)
2020-09-03T16:37:40.7787872Z 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:78)
2020-09-03T16:37:40.7788213Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:132)
2020-09-03T16:37:40.7788546Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-09-03T16:37:40.7788872Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
2020-09-03T16:37:40.7789203Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
2020-09-03T16:37:40.7789530Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
2020-09-03T16:37:40.7789854Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-09-03T16:37:40.7790296Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
2020-09-03T16:37:40.7790779Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
2020-09-03T16:37:40.7791137Z 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
2020-09-03T16:37:40.7791511Z 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
2020-09-03T16:37:40.7791908Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
2020-09-03T16:37:40.7792272Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-09-03T16:37:40.7792637Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
2020-09-03T16:37:40.7792995Z 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
2020-09-03T16:37:40.7793358Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
2020-09-03T16:37:40.7807580Z 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
2020-09-03T16:37:40.7814625Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
2020-09-03T16:37:40.7821736Z 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
2020-09-03T16:37:40.7837009Z 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
2020-09-03T16:37:40.7854379Z 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
2020-09-03T16:37:40.7862095Z 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
2020-09-03T16:37:40.7868006Z 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:248)
2020-09-03T16:37:40.7888823Z 	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$5(DefaultLauncher.java:211)
2020-09-03T16:37:40.7895792Z 	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:226)
2020-09-03T16:37:40.7901840Z 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:199)
2020-09-03T16:37:40.7908297Z 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:132)
2020-09-03T16:37:40.7915646Z 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:188)
2020-09-03T16:37:40.7938889Z 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:154)
2020-09-03T16:37:40.7946648Z 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:128)
2020-09-03T16:37:40.7952961Z 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
2020-09-03T16:37:40.7961434Z 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
2020-09-03T16:37:40.7967015Z 	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
2020-09-03T16:37:40.8004451Z 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
2020-09-03T16:37:40.8009402Z Caused by: java.lang.RuntimeException: Failed to start quarkus
2020-09-03T16:37:40.8014545Z 	at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:146)
2020-09-03T16:37:40.8016444Z 	... 42 more
2020-09-03T16:37:40.8021474Z Caused by: java.lang.RuntimeException: Failed to initialize Arc
2020-09-03T16:37:40.8025655Z 	at io.quarkus.arc.Arc.initialize(Arc.java:26)
2020-09-03T16:37:40.8030438Z 	at io.quarkus.arc.runtime.ArcRecorder.getContainer(ArcRecorder.java:24)
2020-09-03T16:37:40.8039623Z 	at io.quarkus.deployment.steps.ArcProcessor$generateResources-1025303321.deploy_0(ArcProcessor$generateResources-1025303321.zig:76)
2020-09-03T16:37:40.8048346Z 	at io.quarkus.deployment.steps.ArcProcessor$generateResources-1025303321.deploy(ArcProcessor$generateResources-1025303321.zig:40)
2020-09-03T16:37:40.8053480Z 	at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:126)
2020-09-03T16:37:40.8055516Z 	... 42 more
2020-09-03T16:37:40.8062152Z Caused by: java.lang.NoClassDefFoundError: org/eclipse/microprofile/health/HealthCheck
2020-09-03T16:37:40.8067718Z 	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
```

mkouba commented:
@geoand Yep, I'm on it but I have no idea what's going on so far ;-).

mkouba committed:
Disable tests in quarkus-integration-test-vault...

...and quarkus-integration-test-vault-agroal
- see also https://github.com/quarkusio/quarkus/issues/11879
